mender-auth/ipc/server.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-auth/ipc/server.hpp>
+
+#include <functional>
+#include <iostream>
+#include <string>
+
+#include <common/common.hpp>
+#include <common/error.hpp>
+#include <common/events_io.hpp>
+#include <common/http.hpp>
+#include <common/log.hpp>
+
+namespace mender {
+namespace auth {
+namespace ipc {
+
+using namespace std;
+
+const IPCErrorCategoryClass IPCErrorCategory;
+
+const char *IPCErrorCategoryClass::name() const noexcept {
+	return "IPCErrorCategory";
+}
+
+string IPCErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case SetupError:
+		return "Error during setup";
+	case RequestError:
+		return "HTTP client request error";
+	case ResponseError:
+		return "HTTP client response error";
+	case APIError:
+		return "API error";
+	case UnauthorizedError:
+		return "Unauthorized error";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(IPCErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, IPCErrorCategory), msg);
+}
+
+namespace type {
+namespace webserver {
+
+auto NoOpHeaderHandler = [](http::ExpectedIncomingRequestPtr exp_req) {};
+
+// Run a webserver and handle the events as requests
+error::Error Caching::Listen(
+	const string &server_url, const string &private_key_path, const string &identity_script_path) {
+	return this->server_.AsyncServeUrl(
+		server_url,
+		NoOpHeaderHandler,
+		[this, private_key_path, identity_script_path](http::ExpectedIncomingRequestPtr exp_req) {
+			if (!exp_req) {
+				log::Error("Expected request failure: " + exp_req.error().message);
+				return;
+			}
+			auto req = exp_req.value();
```

Comment:
Yeh, I have no idea. Tbh, in cases like this, I'd expect the compiler to be pretty good at understanding where this is used and needed, so that the optimiser handles it in any cases - but might be I'm dreaming...